### PR TITLE
Spark 306189 :: Input.ChoiceSet with markdown URL set in choice "title" appears as raw text.

### PR DESCRIPTION
**Pull request by urmitchauhan**
## Related Issue
The issues fixed with this PR [SPARK-306189](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-306189).

**NOTE**: For multiple issues resolved by this PR use the corresponding keywords **every time** in a comma-delimited list per the reference
page above.

## Description

- Fixed Markdown URL in input choice and input toggle.
- Fixed Wrap

## UI 

Original             	   |  Updated
:-------------------------:|:-------------------------:
image-2022-02-15-17-54-46-319 | Screenshot 2022-05-30 at 3 39 43 PM
Wrap: True Screenshot 2022-05-24 at 12 10 46 PM | Wrap: TrueScreenshot 2022-05-30 at 3 41 44 PM
Wrap: FalseScreenshot 2022-05-24 at 12 14 22 PM | Wrap: False Screenshot 2022-05-30 at 3 48 33 PM

## Sample Card
If appropriate, please include a link to a card in one of the samples directories that can be used to validate this change. This can be an existing card or a card added with this PR.

## How Verified
How you verified the fix, including one or all of the following: 
1. New unit tests that were added if any. If none were added please add a quick line explaining why not.
2. Existing relevant unit/regression tests that you ran
3. Manual scenario verification if any; ***Do include .gif's or screenshots of the testing you performed here if you think that it 
will aid in code reviews or corresponding fixes on other platforms for eg.***

### CODE REVIEW

File: source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm [hunk 1]
```diff
@@ -8,6 +8,18 @@
...
+    auto text = [BridgeConverter getStringCpp:textValue];
+    auto language = [BridgeConverter getStringCpp:@""];
+    std::shared_ptr<MarkDownParser> markDownParser = std::make_shared<MarkDownParser>([BridgeTextUtils getLocalizedDate:text language:language]);
+
+    // MarkDownParser transforms text with MarkDown to a html string
+    NSString *parsedString = [NSString stringWithCString:markDownParser->TransformToHtml().c_str() encoding:NSUTF8StringEncoding];
+    NSData *htmlData = [parsedString dataUsingEncoding:NSUTF16StringEncoding];
+    
+    return [[ACSMarkdownParserResult alloc] initWithParsedString:parsedString htmlData:htmlData];
+}
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm:22]: Add a new line below this

File: source/macos/AdaptiveCards/AdaptiveCards/Renderers/InputToggleRenderer.swift [hunk 2]
```diff
@@ -28,5 +28,25 @@ class InputToggleRenderer: NSObject, BaseCardElementRendererProtocol {
         inputToggleView.labelAttributedString = attributedString
         rootView.addInputHandler(inputToggleView)
         return inputToggleView
-     }
+    }
+    
+    private func getAttributedString(title: String, with hostConfig: ACSHostConfig, renderConfig: RenderConfig, rootView: ACRView, style: ACSContainerStyle) -> NSMutableAttributedString {
+        var attributedString: NSMutableAttributedString
+        attributedString = NSMutableAttributedString(string: title)
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards/Renderers/InputToggleRenderer.swift:35]:
```suggestion
        var attributedString = NSMutableAttributedString(string: title)
```

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift [hunk 3]
```diff
@@ -279,3 +292,30 @@ extension DispatchQueue {
...
+            return super.intrinsicContentSize
+        }
+        layoutManager.ensureLayout(for: textContainer)
+        let size = layoutManager.usedRect(for: textContainer).size
+        let width = size.width + 2
+        return NSSize(width: width, height: size.height)
+    }
+    override func draw(_ dirtyRect: NSRect) {
+        super.draw(dirtyRect)
+    }
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:308]: Does this need override if you just call super in it?

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift [hunk 4]
```diff
@@ -117,7 +128,10 @@ class ACRChoiceButton: NSView, NSTextFieldDelegate, InputHandlingViewProtocol {
         button.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
         button.topAnchor.constraint(equalTo: topAnchor).isActive = true
         button.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
-        buttonLabelField.leadingAnchor.constraint(equalTo: button.trailingAnchor, constant: buttonConfig?.elementSpacing ?? 8).isActive = true
+        button.heightAnchor.constraint(greaterThanOrEqualToConstant: CGFloat(16)).isActive = true
+        button.widthAnchor.constraint(equalToConstant: CGFloat(16)).isActive = true
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:132]: Do you need these 2 even when compressionresistance is set to .required?

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift [hunk 5]
```diff
@@ -117,7 +128,10 @@ class ACRChoiceButton: NSView, NSTextFieldDelegate, InputHandlingViewProtocol {
...
+        button.setContentCompressionResistancePriority(.required, for: .horizontal)
+        buttonLabelField.leadingAnchor.constraint(equalTo: button.trailingAnchor, constant: buttonConfig?.elementSpacing ?? 6).isActive = true
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:134]: Why was the constant changed from 8 to 6?
Any specific reason

File: source/macos/AdaptiveCards/AdaptiveCards/Renderers/InputToggleRenderer.swift [hunk 6]
```diff
@@ -28,5 +28,25 @@ class InputToggleRenderer: NSObject, BaseCardElementRendererProtocol {
...
+        if let colorHex = hostConfig.getForegroundColor(style, color: .default, isSubtle: true), let textColor = ColorUtils.color(from: colorHex) {
+            attributedString.addAttributes([.foregroundColor: textColor], range: NSRange(location: 0, length: attributedString.length))
+        }
+        if attributedString.attributes(at: 0, effectiveRange: nil)[.link] != nil {
+            attributedString.addAttributes([
+                .foregroundColor: renderConfig.hyperlinkColorConfig.foregroundColor,
+                .underlineColor: renderConfig.hyperlinkColorConfig.underlineColor,
+                .underlineStyle: renderConfig.hyperlinkColorConfig.underlineStyle.rawValue
+            ], range: NSRange(location: 0, length: attributedString.length))
+        }
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards/Renderers/InputToggleRenderer.swift:49]: Can you explain this part to me?

File: source/macos/AdaptiveCards/AdaptiveCards/Renderers/ChoiceSetInputRenderer.swift [hunk 7]
```diff
@@ -85,11 +85,13 @@ class ChoiceSetInputRenderer: NSObject, BaseCardElementRendererProtocol {
         return choiceSetFieldCompactView
     }
     
-    private func getAttributedString(title: String, with hostConfig: ACSHostConfig, style: ACSContainerStyle) -> NSMutableAttributedString {
-        let attributedString: NSMutableAttributedString
+    private func getAttributedString(title: String, with hostConfig: ACSHostConfig, renderConfig: RenderConfig, rootView: ACRView, style: ACSContainerStyle) -> NSMutableAttributedString {
+        var attributedString: NSMutableAttributedString
         attributedString = NSMutableAttributedString(string: title)
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards/Renderers/ChoiceSetInputRenderer.swift:90]:
```suggestion
        var attributedString = NSMutableAttributedString(string: title)
```

File: source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm (hunk 1 above)
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm:22]: Ok, will add.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift (hunk 5 above)
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:134]: Constant 8 looks away from radio button, so make it look good by reducing space by 2 px. In addition, we have added 16 constants in width and height.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift (hunk 3 above)
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:308]: nope, not required.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift (hunk 4 above)
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:132]: Yes, without that sometime button is slightly clipped. we have set setContentCompressionResistancePriority for occupancy needed space from right side textview.
Screenshot 2022-05-24 at 1 14 54 PM

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift (hunk 5 above)
Comment (mukuagar) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:134]: Ok sure
Let's just not change in Webex app
As that is according to spec

**Comment by mukuagar** [comment]
The font looks different from screenshots
Any reason for that?

**Comment by urmitchauhan** [comment]
> The font looks different from screenshots Any reason for that?

we need to apply font properties on that.

File: source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm [hunk 8]
```diff
@@ -8,6 +8,19 @@
 
 @implementation BridgeTextUtils
 
++ (ACSMarkdownParserResult * _Nonnull)processOnRawTextString:(NSString * _Nonnull)textValue hostConfig:(ACSHostConfig * _Nonnull)config
+{
+    auto text = [BridgeConverter getStringCpp:textValue];
+    auto language = [BridgeConverter getStringCpp:@""];
```
Comment (sidkotha) [inline source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm:14]: can you add comment in code why we need to pass empty string in method call.

Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm:14]: Yes will add it.

File: source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm [hunk 9]
```diff
@@ -136,4 +137,32 @@ + (ACSRichTextElementProperties * _Nonnull)convertFactToRichTextElementPropertie
     [textProp setLanguage:[fact getLanguage]];
     return textProp;
 }
+
++ (ACSRichTextElementProperties * _Nonnull)convertInputChoiceToRichTextElementProperties:(ACSChoiceInput * _Nonnull)choiceInput
```
Comment (sidkotha) [inline source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm:141]: both these methods have many common lines. can it be extracted to one method.

Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm:141]: Yes, we used direct title text because there is no any other dependancy.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift [hunk 10]
```diff
@@ -11,6 +11,46 @@ protocol ACRChoiceButtonDelegate: NSObjectProtocol {
     func acrChoiceButtonShouldReadError(_ button: ACRChoiceButton) -> Bool
 }
 
+//
+// ACRHyperLinkTextView
+// This class is inherited from NSTextView.
+// This class is used to display a MarkDown Hyperlink in ACRChoiceButton.
+//
+
+class ACRHyperLinkTextView: NSTextView {
```
[the remark this answers is not in the document]
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:20]: create new file for this.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift [hunk 11]
```diff
@@ -117,6 +128,9 @@ class ACRChoiceButton: NSView, NSTextFieldDelegate, InputHandlingViewProtocol {
         button.leadingAnchor.constraint(equalTo: leadingAnchor).isActive = true
         button.topAnchor.constraint(equalTo: topAnchor).isActive = true
         button.bottomAnchor.constraint(equalTo: bottomAnchor).isActive = true
+        button.heightAnchor.constraint(greaterThanOrEqualToConstant: CGFloat(16)).isActive = true
+        button.widthAnchor.constraint(equalToConstant: CGFloat(16)).isActive = true
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:132]: Why equal to for width and >= for height?
Any reason

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRHyperLinkTextView.swift [hunk 12]
```diff
@@ -0,0 +1,44 @@
+//
+//  ACRHyperLinkTextView.swift
+//  AdaptiveCards
+//  This class is inherited from NSTextView.
+//  This class is used to display a MarkDown Hyperlink in ACRChoiceButton.
+//  Created by uchauhan on 30/05/22.
```
Comment (mukuagar) [inline source/macos/AdaptiveCards/AdaptiveCards/Views/ACRHyperLinkTextView.swift:6]: Great job adding this!

**Review by mukuagar — COMMENTED** [review]
Minor comments anddiscussions
Looks good!

File: source/macos/AdaptiveCards/AdaptiveCards/Renderers/ChoiceSetInputRenderer.swift [hunk 13]
```diff
@@ -85,11 +85,14 @@ class ChoiceSetInputRenderer: NSObject, BaseCardElementRendererProtocol {
...
-        attributedString = NSMutableAttributedString(string: title)
+    private func getAttributedString(title: String, with hostConfig: ACSHostConfig, renderConfig: RenderConfig, rootView: ACRView, style: ACSContainerStyle) -> NSMutableAttributedString {
+        let markdownResult = BridgeTextUtils.process(onRawTextString: title, hostConfig: hostConfig)
+        let markdownString = TextUtils.getMarkdownString(for: rootView, with: markdownResult)
+        let textProperties = BridgeTextUtils.getRichTextElementProperties(title)
+        let attributedString = TextUtils.addFontProperties(attributedString: markdownString, textProperties: textProperties, hostConfig: hostConfig)
+        
         if let colorHex = hostConfig.getForegroundColor(style, color: .default, isSubtle: true), let textColor = ColorUtils.color(from: colorHex) {
-            attributedString.addAttributes([.foregroundColor: textColor], range: NSRange(location: 0, length: attributedString.length))
+            attributedString.addAttributes([.foregroundColor: textColor, .cursor: NSCursor.arrow], range: NSRange(location: 0, length: attributedString.length))
```
[the remark this answers is not in the document]
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards/Renderers/ChoiceSetInputRenderer.swift:95]: When we set the cursor to be an arrow, the user experience is similar to that of a button element, and the user cannot easily select a word as the ibeam cursor. still user can select the text from that. there some limitation.

File: source/macos/AdaptiveCards/AdaptiveCards/Renderers/InputToggleRenderer.swift [hunk 14]
```diff
@@ -28,5 +28,17 @@ class InputToggleRenderer: NSObject, BaseCardElementRendererProtocol {
...
+        let markdownResult = BridgeTextUtils.process(onRawTextString: title, hostConfig: hostConfig)
+        let markdownString = TextUtils.getMarkdownString(for: rootView, with: markdownResult)
+        let textProperties = BridgeTextUtils.getRichTextElementProperties(title)
+        let attributedString = TextUtils.addFontProperties(attributedString: markdownString, textProperties: textProperties, hostConfig: hostConfig)
+        
+        if let colorHex = hostConfig.getForegroundColor(style, color: .default, isSubtle: true), let textColor = ColorUtils.color(from: colorHex) {
+            attributedString.addAttributes([.foregroundColor: textColor, .cursor: NSCursor.arrow], range: NSRange(location: 0, length: attributedString.length))
+        }
+        return attributedString
+    }
```
[the remark this answers is not in the document]
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards/Renderers/InputToggleRenderer.swift:43]: you mean make separate  function in textutils. yes we can that, need to check.

File: source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm [hunk 15]
```diff
@@ -123,4 +137,18 @@ + (ACSRichTextElementProperties * _Nonnull)convertFactToRichTextElementPropertie
     [textProp setLanguage:[fact getLanguage]];
     return textProp;
 }
+
++ (ACSRichTextElementProperties * _Nonnull)getRichTextElementProperties:(NSString * _Nonnull)elementText
+{
+    ACSRichTextElementProperties* textProp = [[ACSRichTextElementProperties alloc] initWithRichTextElementProperties:std::make_shared<RichTextElementProperties>()];
+    [textProp setText: elementText];
+    [textProp setTextSize:ACSTextSizeSmall];
```
[the remark this answers is not in the document]
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards-bridge/BridgeUtils/BridgeTextUtils.mm:145]: default font is look like bigger in this. will catch up for that.

File: source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift (hunk 11 above)
Comment (urmitchauhan) [reply source/macos/AdaptiveCards/AdaptiveCards/Views/ACRChoiceButton.swift:132]: Yes, we only need the width equivalence constant because it does not shrink where we set it, and height may have an impact on other elements of the stack. We require a minimum of some constant height that is not less than that. where We need the exact width, not more or less. my intentions is to not affect when there is multiline or other things in stack. let's test some edge case.
